cpp/src/replace/replace.cu

```diff
@@ -53,16 +53,27 @@ namespace{ //anonymous
                       gdf_size_type                      nrows,
                       thrust::device_ptr<const T> old_values_begin,
                       thrust::device_ptr<const T> old_values_end,
-                      const T*                    d_new_values)
+                      const T*                    d_new_values,
+                      gdf_valid_type*             col_valid,
+                      gdf_valid_type*             new_valid)
   {
     gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
     while(i < nrows)
     {
-      auto found_ptr = thrust::find(thrust::seq, old_values_begin, old_values_end, d_col_data[i]);
-
-      if (found_ptr != old_values_end) {
-          auto d = thrust::distance(old_values_begin, found_ptr);
-          d_col_data[i] = d_new_values[d];
+      if ( gdf_is_valid(col_valid, i)){
+          auto found_ptr = thrust::find(thrust::seq, old_values_begin, old_values_end, d_col_data[i]);
+
+          if (found_ptr != old_values_end) {
+              auto d = thrust::distance(old_values_begin, found_ptr);
+              if (gdf_is_valid(new_valid, d)){
```

Comment:
For performance, use bit_mask::is_valid() with the bit_mask_t pointer.

Reply:
Also, I think this function checks whether the valid pointer is null, which is expensive to do for every thread. is_valid() doesn't, so you will need to check outside the kernel and then have a bool template parameter on the kernel for  whether each valid pointer is null or not. Then in your dispatcher you need to if/else over the 4 combinations and dispatch the kernel with the appropriate template parameters. 

See #1654

Reply:
PR #1654 really did help me to use bit_mask utility.